### ISSUE TITLE
Frequent failures in odo service command tests for OperatorHub

### ISSUE THREAD

**Issue by dharmit** [openshift/odo#3193]
/kind failing-test

[odo test statistics dashboard](https://gist.github.com/jgwest/5b28687514970ae6baa5942ee82d5755) is reporting quite a few failures for `odo service command tests for OperatorHub When operator backed services are created`.

When we delete `example3` service using `oc delete EtcdCluster/example3`, it succeeds. Next we check if `odo service list` returns the message `No operator backed services found in the namesapce` but it doesn't. From the logs, it seems like `example2` service started before `example3` test is not getting deleted.

**Comment by dharmit** [comment]
Surprisingly, I've never seen this issue in local environment when I run the tests. It's only showing up on the CI. Looks like something related to parallel tests running on multiple CI nodes.

**Comment by dharmit** [comment]
> Surprisingly, I've never seen this issue in local environment when I run the tests. It's only showing up on the CI. Looks like something related to parallel tests running on multiple CI nodes.

@amitkrout can you please help understand what's the problem here? I have a feeling that we'll need to use `helper.WaitForCmdOut` so that this issue doesn't pop up. Or, change the expectation for `stdOut` to not contain the name of the service being deleted (here `example3`) instead of expecting it to contain `No operator backed services found in the namesapce`.

**Comment by dharmit** [comment]
Here's the error that it keeps showing:
```
Running oc with args [oc delete EtcdCluster example3]
[oc] etcdcluster.etcd.database.coreos.com "example3" deleted
Running odo with args [odo service list]
[odo] I0522 05:29:24.393559    9641 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:24.393690    9641 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:24.510108    9641 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:24.510230    9641 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:24.510295    9641 service.go:227] Getting list of services
[odo] I0522 05:29:24.510331    9641 operators.go:17] Fetching list of operators installed in cluster
[odo] I0522 05:29:24.602114    9641 service.go:239] Getting services started from operator: etcdoperator.v0.9.4
[odo] I0522 05:29:24.602158    9641 service.go:277] Getting instances of: etcdclusters.etcd.database.coreos.com
[odo] I0522 05:29:24.632337    9641 service.go:277] Getting instances of: etcdbackups.etcd.database.coreos.com
[odo] I0522 05:29:24.659886    9641 service.go:277] Getting instances of: etcdrestores.etcd.database.coreos.com
[odo] I0522 05:29:24.687668    9641 service.go:239] Getting services started from operator: mongodb-enterprise.v1.5.2
[odo] I0522 05:29:24.687698    9641 service.go:277] Getting instances of: mongodb.mongodb.com
[odo] I0522 05:29:24.715236    9641 service.go:277] Getting instances of: mongodbusers.mongodb.com
[odo] I0522 05:29:24.742976    9641 service.go:277] Getting instances of: opsmanagers.mongodb.com
[odo] NAME         TYPE            AGE
[odo] example2     EtcdCluster     14s


[91m[1m• Failure [18.765 seconds][0m
odo service command tests for OperatorHub
[90m/go/src/github.com/openshift/odo/tests/integration/operatorhub/cmd_service_test.go:21[0m
  When operator backed services are created
  [90m/go/src/github.com/openshift/odo/tests/integration/operatorhub/cmd_service_test.go:169[0m
    [91m[1mshould list the services if they exist [It][0m
    [90m/go/src/github.com/openshift/odo/tests/integration/operatorhub/cmd_service_test.go:170[0m

    [91mFailed after 0.498s.
    Running odo with args [odo service list]
    Expected
        <int>: 0
    not to match exit code:
        <int>: 0[0m
```
Link to one of the [failed tests](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_odo/3224/pull-ci-openshift-odo-master-v4.2-integration-e2e/592/build-log.txt).

**Comment by dharmit** [comment]
Due to the parallel execution of the tests, it seems to me that last test in the file is executed first:
```
+ make test-operator-hub
ginkgo  -randomizeAllSpecs -slowSpecThreshold=120 -timeout 7200s -nodes=2 -focus="odo service command tests" tests/integration/operatorhub/
Running Suite: Operatorhub Suite
================================
Random Seed: [1m1590125344[0m - Will randomize all specs
Will run [1m7[0m specs

Running in parallel across [1m2[0m nodes

[32m•[0m[32m•[0m[32m•[0m[32m•[0m
[90m------------------------------[0m
Running odo with args [odo project set ci-operator-hub-project]
[odo] Switched to project : ci-operator-hub-project
Running odo with args [odo catalog list services]
[odo] I0522 05:29:06.419040    8848 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:06.540933    8848 operators.go:17] Fetching list of operators installed in cluster
[odo] I0522 05:29:06.652839    8848 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:06.652966    8848 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] Operators available in the cluster
[odo] NAME                          CRDs
[odo] etcdoperator.v0.9.4           EtcdCluster, EtcdBackup, EtcdRestore
[odo] mongodb-enterprise.v1.5.2     MongoDB, MongoDBUser, MongoDBOpsManager
Running odo with args [odo service create etcdoperator.v0.9.4 --crd EtcdCluster --dry-run]
[odo] I0522 05:29:06.877562    8878 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:06.877740    8878 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:06.877798    8878 operators.go:17] Fetching list of operators installed in cluster
[odo] I0522 05:29:06.968051    8878 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:06.968185    8878 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] apiVersion: etcd.database.coreos.com/v1beta2
[odo] kind: EtcdCluster
[odo] metadata:
[odo]   name: example
[odo] spec:
[odo]   size: 3
[odo]   version: 3.2.13
[odo] 
Running odo with args [odo service create --from-file /tmp/anrgha.yaml]
[odo] I0522 05:29:07.189731    8918 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:07.189949    8918 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:07.190154    8918 operators.go:17] Fetching list of operators installed in cluster
[odo] I0522 05:29:07.284867    8918 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] Deploying service of type: EtcdCluster
[odo]  •  Deploying service  ...
[odo] I0522 05:29:07.285103    8918 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] 
 ✓  Deploying service [46ms]
[odo]  ✓  Service 'example3' was created
[odo] 
[odo] Progress of the provisioning will not be reported and might take a long time
[odo] You can see the current status by executing 'odo service list'
[odo] Optionally, link  to your component by running: 'odo link <component-name>'
Running oc with args [oc get pods -n ci-operator-hub-project]
[oc] NAME                             READY   STATUS     RESTARTS   AGE
[oc] etcd-operator-7bf4c7ddb8-msjmf   3/3     Running    0          77m
[oc] example3-wrhp5mb6nx              0/1     Init:0/1   0          0s
Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Pending"Running oc with args [oc get pods example3-wrhp5mb6nx -o template="{{.status.phase}}" -n ci-operator-hub-project]
[oc] "Running"Running odo with args [odo service list]
[odo] I0522 05:29:23.020845    9574 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:23.021014    9574 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:23.141125    9574 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:23.141287    9574 preference.go:188] The path for preference file is /tmp/artifacts/.odo/preference.yaml
[odo] I0522 05:29:23.141379    9574 service.go:227] Getting list of services
[odo] I0522 05:29:23.141392    9574 operators.go:17] Fetching list of operators installed in cluster
[odo] I0522 05:29:23.229627    9574 service.go:239] Getting services started from operator: etcdoperator.v0.9.4
[odo] I0522 05:29:23.229662    9574 service.go:277] Getting instances of: etcdclusters.etcd.database.coreos.com
[odo] I0522 05:29:23.257473    9574 service.go:277] Getting instances of: etcdbackups.etcd.database.coreos.com
[odo] I0522 05:29:23.284714    9574 service.go:277] Getting instances of: etcdrestores.etcd.database.coreos.com
[odo] I0522 05:29:23.312516    9574 service.go:239] Getting services started from operator: mongodb-enterprise.v1.5.2
[odo] I0522 05:29:23.312547    9574 service.go:277] Getting instances of: mongodb.mongodb.com
[odo] I0522 05:29:23.339807    9574 service.go:277] Getting instances of: mongodbusers.mongodb.com
[odo] I0522 05:29:23.368596    9574 service.go:277] Getting instances of: opsmanagers.mongodb.com
[odo] NAME         TYPE            AGE
[odo] example2     EtcdCluster     13s
[odo] example3     EtcdCluster     16s
```
In above output, age of `example2` service is less than that of `example3` service. As a result, this particular check is going to fail: https://github.com/openshift/odo/blob/fb8315105361c8c5accb72713685e8e1a4d722d2/tests/integration/operatorhub/cmd_service_test.go#L222-L223

I'll open a PR with that part modified. But it would mean that we can't check for the following two outputs in tests:
```sh
$ odo service list
odo service list            
 ✗  No operator backed services found in the namesapce
```
and:
```sh
odo service list -o json
{
    "kind": "Error",
    "apiVersion": "odo.dev/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "No operator backed services found in the namesapce"
}
```

**Comment by prietyc123** [comment]
> I'll open a PR with that part modified. But it would mean that we can't check for the following two outputs in tests:

TBH I don't understand what is the use case you are trying to resolve here. Can you please add some more context?

> $ odo service list
odo service list            
 ✗  No operator backed services found in the namesapce

I am confused in which namespace operator backed service is not found. Ideally we should display the name of the `namespace` as well for better user experience. I think we can create a separate pr to update it.

**Comment by dharmit** [comment]
> TBH I don't understand what is the use case you are trying to resolve here. Can you please add some more context?

Honestly, I'm not sure how to add any more context than what I have already added. It's a test that's failing on CI but not in my local setup. There's also this comment https://github.com/openshift/odo/issues/3193#issuecomment-629956794. And the test in question is: https://github.com/openshift/odo/blob/fb8315105361c8c5accb72713685e8e1a4d722d2/tests/integration/operatorhub/cmd_service_test.go#L169-L170

I should have mentioned the test that's failing, my bad.

> > $ odo service list
> > odo service list
> > ✗  No operator backed services found in the namesapce
> 
> I am confused in which namespace operator backed service is not found. Ideally we should display the name of the `namespace` as well for better user experience. I think we can create a separate pr to update it.

I'm not sure why the user experience matters when we're discussing the tests. But anyway, since `odo service list` in the experimental mode doesn't accept extra flags like `--app`, `--project`, we've ensured that proper namespace is set before each spec is executed. https://github.com/openshift/odo/blob/fb8315105361c8c5accb72713685e8e1a4d722d2/tests/integration/operatorhub/cmd_service_test.go#L23-L29

I think I know how to fix this test from failing. I'll open a PR and we'll hopefully be able to address it. That should also help clear the confusion.

**Comment by prietyc123** [comment]
> When we delete example3 service using oc delete EtcdCluster/example3, it succeeds. Next we check if odo service list returns the message No operator backed services found in the namesapce but it doesn't. From the logs, it seems like example2 service started before example3 test is not getting deleted.

Yes, This will happen when we are running parallel test in a single [namespace](https://github.com/openshift/odo/blob/master/tests/integration/operatorhub/cmd_service_test.go#L26) . 

So the possible solution would be run it sequentially to avoid the flake that you ar facing right now or create separate namespace for each operator backed service and define it during the test.

**Comment by prietyc123** [comment]
> It's a test that's failing on CI but not in my local setup. 

Yes, sometimes CI behaves slowly which causing the issue. As part of your local setup is concerned may be you have a goood  set up configuration to run this test. Hence there is a less chance of getting such Flake.

> I'm not sure why the user experience matters when we're discussing the tests. 

We are hiding the information from the user like which namespace in which namespace operator backed service is not found. This is not about the test, I am talking in general. So adding the namespace name in the message will add more weightage. 

> But anyway, since odo service list in the experimental mode doesn't accept extra flags like --app, --project. 

I checked while debugging the failure and found that we do have flags like --app, --project, even in experimental mode.

```
$ odo preference view
PARAMETER             CURRENT_VALUE
UpdateNotification    false
NamePrefix            
Timeout               
PushTimeout           
Experimental          true
PushTarget            
$ odo service list -h
List all services in the current application

[...]
Flags:
      --app string       Application, defaults to active application
      --context string   Use given context directory as a source for component settings
  -h, --help             Help for list
      --project string   Project, defaults to active project
[...]
```

> we've ensured that proper namespace is set before each spec is executed.

Yes, thats true even I can see in the [test](https://github.com/openshift/odo/blob/master/tests/integration/operatorhub/cmd_service_test.go#L26).

**Comment by amitkrout** [comment]
> I have a feeling that we'll need to use helper.WaitForCmdOut so that this issue doesn't pop up. 

I am sure this won't work in parallel execution because you never know the immediate parallel test may add more similar result to the same command you trying to verify. 

> Or, change the expectation for stdOut to not contain the name of the service being deleted (here example3) instead of expecting it to contain No operator backed services found in the namesapce.

This will be a good approach if we want to stick to parallel run. However @prietyc123 suggested series run also deserves a mention.

> Due to the parallel execution of the tests, it seems to me that last test in the file is executed first

yes, ginkgo uses a random algorithm to pick test in parallel run. So this is expected.

**Comment by amitkrout** [comment]
> I'll open a PR with that part modified. But it would mean that we can't check for the following two outputs in tests:
$ odo service list
odo service list            
 ✗  No operator backed services found in the namesapce
and:
odo service list -o json
{
    "kind": "Error",
    "apiVersion": "odo.dev/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "No operator backed services found in the namesapce"
}

Let's not compromise test spec due to parallel run constraint on a single namespace. Then i guess @prietyc123 has nice suggestion of running test in series.

**Comment by dharmit** [comment]
> > 
> 
> I checked while debugging the failure and found that we do have flags like --app, --project, even in experimental mode.
> 
> ```
> 
> ```

They are available doesn't translate to they are working. :wink: 

Anyway, in this case `odo service list --project myproject` works. But, if you use `--app` flag it would fail till #3134 is merged and I use that to take #2801 to a closure.

The relief in this case was the my assumption about `--project` not working was incorrect. Since it's working, I've modified the tests without any compromise to the spec and making it less ugly than what I had come up with earlier. :smile:

**Comment by prietyc123** [comment]
> > I checked while debugging the failure and found that we do have flags like --app, --project, even in experimental mode.
> 
> They are available doesn't translate to they are working. 😉
> 
> Anyway, in this case `odo service list --project myproject` works. But, if you use `--app` flag it would fail till #3134 is merged and I use that to take #2801 to a closure.
> 
> The relief in this case was the my assumption about `--project` not working was incorrect. Since it's working, I've modified the tests without any compromise to the spec and making it less ugly than what I had come up with earlier. 😄

Thanks @dharmit for detailed information. Actually I was really worried if these flags are not working then why are they still visible in UX experimental mode. I was about to create the issue on the basis of my observation. But thanks to clear my doubts.
